crates/turborepo-lib/src/cli.rs

```diff
@@ -202,8 +203,44 @@ pub enum LinkTarget {
 
 impl Args {
     pub fn new() -> Result<Self> {
-        let mut clap_args = match Args::try_parse() {
-            Ok(args) => args,
+        // We always pass --single-package in from the shim.
+        // We need to omit it, and then add it in for run.
+        let arg_separator_position = std::env::args_os()
+            .find_position(|input_token| input_token == "--")
+            .map(|(index, _)| index);
+
+        let single_package_position = std::env::args_os()
+            .find_position(|input_token| input_token == "--single-package")
+            .map(|(index, _)| index);
+
+        let is_single_package = match (arg_separator_position, single_package_position) {
+            (_, None) => false,
+            (None, Some(_)) => true,
+            (Some(arg_separator_position), Some(single_package_position)) => {
+                single_package_position < arg_separator_position
+            }
+        };
+
+        // Clap supports arbitrary iterators as input.
+        // We can remove all instances of --single-package
+        let single_package_free = std::env::args_os()
+            .enumerate()
+            .filter(|(index, input_token)| {
+                arg_separator_position
+                    .is_some_and(|arg_separator_position| index > &arg_separator_position)
+                    || input_token != "--single-package"
+            })
+            .map(|(_, input_token)| input_token);
+
+        let mut clap_args = match Args::try_parse_from(single_package_free) {
+            Ok(mut args) => {
+                // And then only add them back in when we're in `run`.
+                if let Some(ref mut run_args) = args.run_args {
```

Comment:
Should we only add it if we've detected a single package flag?

Reply:
That's what the next line does:
```
run_args.single_package = is_single_package
```

This is a computed value, based upon the presence of the flag.

https://github.com/vercel/turbo/pull/6086/files#diff-ff97a6d3bfd4f86fbdd5a7ef844ba82b0371fdd6e016919429b4ba63e725f7a3R216-R222

And, since we're looking for `Some(run_args)`, it ends up applied only to the `run` command.

Reply:
Ah gotcha, nvm then